Fix #5388. Made Vector2 angle methods consistent

obigu opened the pull request:
Fix #5388 and #5385.

This PR contains two different changes to Vector2 angle methods:
1. Fixed `angle(Vector2)` and `angleRad(Vector2)` to return counter-clockwise angles as specified in their javadocs.
2. Made `angle(Vector2)` consistent to `angle()` to return values in the [0, 360) range.

obigu commented:
Well, there are 2 different points on the PR.

Point **1.** is a matter of fixing a wrong implementation which I think comes as top priority over backwards compatibility (otherwise the only alternative solution is to change the javadocs to reflect what it actually does, mark it as deprecated and create a new one that is consistent with the rest of methods which are counter-clockwise). Or do we just ignore it's wrong?

Regarding **2.** it does break backwards compatibility and it may be argued that for not a good enough reason, but not "heavily". There will be some cases in which it does break something but it won't in the majority of cases as angles are usually dealt with no matter the range. 

**If the PR is not being merged due to point **2.** I'm happy to remove it and keep only **1.****

ShibaBandit commented:
I would recommend just clarifying JavaDocs to the actual behavior.

obigu commented:
> I would recommend just clarifying JavaDocs to the actual behavior.

I assume you've read https://github.com/libgdx/libgdx/issues/5388. 

If we **just** change the javadocs (and do not deprecate tham and create new consistent ones) we would be validating/embracing a bad API and giving up on improving it. 

Both approaches (breaking backwards compatibility or deprecation) are valid and acceptable depending on the case, taking shortcuts isn't.

crykn commented:
Hm, my two cents on this issue: I thinks, in general, it makes sense to fix to fix these inconsistencies. However, to make it more obvious that something changed with the angle methods, I vote for renaming them to angleDeg(), as suggested in one of the initial issues. This way, the overall naming is more consistent and only `angleRad(Vector2)` returns different values than before.

obigu commented:
I think it's a good suggestion @crykn. By renaming them to `angleDeg`, not only we "force" the users to look for the new method in CHANGES but we also make the name more meaningful (I've found myself checking in docs if `angle()` returned `deg` or `rad` more than once). I'll update the PR.

Darkyenus commented:
How about keeping the old methods and deprecating them for a few versions (then delete)? The price is very low and it would be a good opportunity to update the javadocs of the old methods to mention that the replacements have a different behavior.

yuripourre commented:
@Darkyenus now that libgdx is going to have more frequent releases sounds a very good idea, indeed.

obigu commented:
@crykn @Darkyenus I've added new methods with "deg" suffix, deprecated the old ones and added javadocs were required. To make it consistent I've done it for all methods (not only `angle()` and `angle(Vector2)`. The only backwards compatibility change affects `angleRad(Vector2)` (which should not be a very commonly used method) and CHANGES have been updated to mention it especifically.

crykn reviewed (approved):
I don't think deprecating `rotate90 (int dir)` is needed, but I don't mind.

obigu commented:
> I don't think deprecating `rotate90 (int dir)` is needed, but I don't mind.

It's just an unnecessary method

Darkyenus commented:
I am also against deprecating `rotate90`. While it is a functional subset, it is computationally more efficient, and, more importantly, 100% precise.

obigu commented:
I don't agree with your point, I don't think it makes sense to accept this extra computational cost and lack of precision for any other rotation except for 90º (if that's critical for an app it should implement higher precision logic on their end). But from the thumbs up votes I guess most of you are in favout of keeping it so I go with the majority.

My question before making the change is, if this method is useful for the API, shall we create the Radians version (that rotates PI/2) for consistency?

Darkyenus commented:
Why break existing code over something that works well? I would keep the name as it is, 90 in context of rotations is self explanatory.

obigu commented:
Requested change made

mgsx-dev commented on gdx/src/com/badlogic/gdx/math/Vector2.java:
```diff
@@ -385,6 +415,22 @@ public Vector2 rotateRad (float radians) {
 		return this;
 	}
 
+	/** Rotates the Vector2 by the given angle around reference vector, counter-clockwise assuming the y-axis points up.
```
could you please don't change methods order, it messing the diff and history, it looks like you added a new method rotateAround instead of modifying it.

obigu replied:
the change of order on `rotateAround` was deliberate to try to keep the 3 versions of each method together for clarity but that's ok, I'll revert.

mgsx-dev commented on gdx/src/com/badlogic/gdx/math/Vector2.java:
```diff
@@ -316,19 +316,40 @@ public float crs (float x, float y) {
 		return this.x * y - this.y * x;
 	}
 
+
```
you can remove empty line (or use formater on modified blocs)

obigu replied:
thanks, missed that one. Unfortunately using formatter is still not an option, last time I ran it on modified code I had to change it all over again due to format differences between IDEs/formatters.
